### ISSUE TITLE
Increase performance of fixed effects demeaning

### ISSUE THREAD

**Issue by s3alfisc** [s3alfisc/pyfixest#98]
It is very, very slow. 

Current implementation: 

```py
        for x in interacted_fes:
            vars = x.split("^")
            data[x] = data[vars].apply(lambda x: '^'.join(
                x.dropna().astype(str)) if x.notna().all() else np.nan, axis=1)

```
See [here](https://github.com/s3alfisc/pyfixest/blob/533675434bfc8cdcda08b60471165243554fbe84/pyfixest/fixest.py#L164).

**Comment by grantmcdermott** [comment]
Hi Alex,

I (finally) had a chance to test drive pyfixest recently. I was going to collect a few comments, but the first thing that jumped out at me is that it a significantly slower than r-fixest. I have a set of standing benchmarks that I use for comparing software (over a whole bunch of tasks, but HFDE regression is one of them). For this particular task, I use a 3-month subset of NYC taxi data, since I prefer to use "real-life" data where possible. 

R (skipping the data prep stuff)

```r
tic = Sys.time()
invisible(feols_reg = feols(
        tip_amount ~ trip_distance + passenger_count |
            dofw + vendor_id + payment_type,
        data = nyc, vcov = "iid"
))
(toc = Sys.time() - tic)
# Time difference of 12.35961 secs
```

Python (ditto)

```py
%timeit
feols_reg = feols(
    fml="tip_amount ~ trip_distance + passenger_count | dofw + vendor_id + payment_type", 
    data=nyc, vcov="iid"
)
# 4m 28.3s
```

As you can see, there's 20x difference. (Am I doing something wrong w.r.t. syntax? I don't think so and the coef results are identical.)

Sorry, I don't mean to complain. There's loads to like here and I really think you've done a fantastic job of mimicking fixest's API within the contraints of Python syntax. But you did ask for feedback, so here it is!

**Comment by s3alfisc** [comment]
Hi Grant, 

First of all, thanks for the feedback, and also the nice words! I'm always relieved to hear that `fixest` and `pyfixest` produce identical results on fresh data sets 😄 

I don't think that you're doing anything wrong - `PyFixest` is just much slower than `fixest` at the moment. So far, my development philosophy has been to "implement the most important features fast and ignore performance", so speeding things up really hasn't been that much of a concern in contrast to ease of use. But now that most features are (more or less) implemented, I really should start to tackle performance. 20x is indeed not great. 

In principle, there are two important bottlenecks: I might have implemented some sub-optimal routines in the internal data preprocessing, and then there is the actual demeaning algorithm. 

On data processing, I just did some quick profiling and indeed, changing a few lines here and there does already help a bit. Another option is to move the back end from `pandas` to `polars`, but I am not too keen to do so, mostly because `formulaic` returns `pd.DataFrame` out of the box, and because I have (strangely) started to find `pandas` indices to be rather useful 😅 

But only around 15% of runtime is spent in the `model_matrix` part, so I see somewhat limited optimization potential here. Though this might be worse with your data set? 

The other big chunk (around 80%) is spent demeaning the hdfe's. Here, `pyfixest` builds on `pyhdfe`, which implements the demeaning algo in pure `numpy`, without parallelism and - in the version that I am using here - without any acceleration. 

I've tried to implement a faster algo in numba / cpp a while back, but did not really succeed to beat `pyhdfe` in terms of speed. We've had some [loose conversations ](https://github.com/s3alfisc/pyfixest/issues/66) in which I tried to groom @lrberge and @pachadotdev to help me port the fixest cpp code. I definitely need to get up to speed on how to integrate c++ code into python packages - unfortunately, it looks to me like it is a little bit more complex than in R vic rcpp.

**Comment by pachadotdev** [comment]
Hi

I can help to write in pure C++. I am taking courses about it

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by lrberge** [comment]
Hi everyone. As we discussed, I should be able to write a pure C++ API of the demeaning algo. But I need to get out of my teaching crunch!

**Comment by pachadotdev** [comment]
That would be amazing! Are you ok if I write a skeleton in the meanwhile?

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by lrberge** [comment]
Thanks! But that should be all right. I have a very precise idea of the implementation in my head already.

**Comment by s3alfisc** [comment]
Amazing @lrberge! I'll make sure to get up to speed on Python c++ bindings in the meantime & will also reach out to @jeffgortmaker to see if he has a preferred way of doing things, as I would then try to work towards a PR for pyhdfe. 😃

**Comment by grantmcdermott** [comment]
> Another option is to move the back end from pandas to polars, but I am not too keen to do so, mostly because formulaic returns pd.DataFrame out of the box, and because I have (strangely) started to find pandas indices to be rather useful 😅

Ha, polars support was another request. To each their own (and it the ubiquity of use makes it the sensible default), but I thoroughly dislike the pandas API despite many years spent _trying_ to like it. There are many reasons and it always feels like a clunky step backwards from any of the equivalent R data implementations. But let's just say that indices are a prime culprit ;-)

At anyway rate, I certainly shouldn't distract with a pandas VS polars kvetch here. Sounds like you have a good roadmap for improved performance over the medium-term. Thanks again for taking my comments in the constructive spirit that they were intended, say nothing of the other great FOSS contributions that you've been putting out.

**Comment by s3alfisc** [comment]
Hi @grantmcdermott , some crazy things happened over the last two days, which lead to the release of version `0.10`. Motivated by your comment, I posted an old `numba` implementation of the demeaning algo (that I had put together a while back) in the [numba forum ](https://numba.discourse.group/t/optimization-of-numba-implementation-of-an-alternating-projection-algorithm/2193), hoping for some help. And to my big surprise, I indeed got it! @styfenschaer responded to my inquiry, and less than 12h later, `PyFixest` is around 5x faster! 

Here's a benchmark of the "old" demeaning algo in contrast to Styven's implementation (on my admittedly old pc): 

```py
%load_ext autoreload
%autoreload 2

import numpy as np
import time
import pyhdfe
from pyfixest.demean import demean

np.random.seed(1238)
N = 10_000_000
x = np.random.normal(0, 1, 10*N).reshape((N,10))
f1 = np.random.choice(list(range(1000)), N).reshape((N,1))
f2 = np.random.choice(list(range(1000)), N).reshape((N,1))

flist = np.concatenate((f1, f2), axis = 1)
weights = np.ones(N)

algorithm = pyhdfe.create(flist)

start_time = time.time()
res_pyhdfe = algorithm.residualize(x)
end_time = time.time()
print(end_time - start_time)
# 26.04527711868286


start_time = time.time()
res_pyfixest, success = demean(x, flist, weights, tol = 1e-10)
# Calculate the execution time
end_time = time.time()
print(end_time - start_time)
#4.334428071975708

np.allclose(res_pyhdfe , res_pyfixest)
# True
```
Sometimes, OSS development is just super cool. 🥳 Thanks for pushing me in the right direction, and special thanks to @styfenschaer for helping me with this!

**Comment by grantmcdermott** [comment]
> Sometimes, OSS development is just super cool. 🥳 Thanks for pushing me in the right direction, and special thanks to @styfenschaer for helping me with this!

Hear! hear!

I just updated to 0.10 and the MWE above works great... Unfortunately, and I apologize for pouring some cold water on progress here, but the 0.10 update has had the opposite effect for my actual use case. Specifically, I believe it may have introduced a convergence instability, since I don't get a result even after 45 minutes. I terminated my kernel at this point and rebooted my computer to make sure. But a second test was dragging passed 15 minutes, so I cancelled that one too.

I haven't changed my code at all, but perhaps it's something weird/dumb on my end. I'll post a separate issue with my full code so you can take a look.

**Comment by styfenschaer** [comment]
Hi @s3alfisc 

I had a quick look at how `pyhdfe` checks convergence and it think they use a softer criterion:
```python
def max_norm_convergence(last_matrix: Array, matrix: Array, tol: float) -> bool:
    """Check for max norm convergence."""
    return np.abs(last_matrix - matrix).max() < tol
```
Compared to what `pyfixest` now does (this is a column-wise check):
```python
@nb.njit
def _sad_converged(a, b, tol):
    for i in range(a.size):
        tol -= np.abs(a[i] - b[i])
        if tol < 0:
            return False
    return True
```

The quivalent to `pyhdfe` would be: 
```python
@nb.njit
def _sad_converged(a, b, tol):
    for i in range(a.size):
        if np.abs(a[i] - b[i]) >= tol:
            return False
    return True
```

The larger the dataset, the larger the difference between the two approaches. However, you have a better understanding of this matter. What do you think?

**Comment by s3alfisc** [comment]
Thanks @styfenschaer! For reasons of alignment with pyhdfe, I think I'd be in favor of moving to the second option =) 

Though I'm not sure if this will solve @grantmcdermott's convergence issues. I was thinking that maybe they are related to problems with jit compilation with a very large initial problem? This is something I have experienced with Julia's jit compiler - if the compilation happens for a complex problem, it might take significantly longer to compile a piece of code. Maybe this applies to numba as well?

**Comment by s3alfisc** [comment]
For reference: 
- [fixest convergence criterion ](https://github.com/lrberge/fixest/blob/a4d1a9bea20aa7ab7ab0e0f1d2047d8097971ad7/src/demeaning.cpp#L84) (default tolerance 1e-06)
- [alpaca convergence criterion](https://github.com/amrei-stammann/alpaca/blob/c9ce131d949327e8b261f1df9a7d02823c5343ff/src/CenterVariables.cpp#L66)

**Comment by s3alfisc** [comment]
Here are some timings of the nyc taxi data set above. Overall, the entire model is fit in around 2 minutes using version `0.10.3`.

- FixestMulti init: `1.84s`- this is due to `self._data.copy()`, which is costly with large data frames. I could offer an option not to copy the data frame internally, which should speed things up, but is less safe
- FixestMulti prep: `0.0` - negligible
- model_matrix took `1.49` seconds - very fast
- model_matrix_fixest: `6.94` seconds-  I can probably speed things up by creating a `numpy` model matrix instead of a `pd.DataFrame` and use formulaic model info to access indices. potentially, moving to `polars` might also be beneficial
- Time to create algorithm: `95.7` seconds - currently, the package only initiates `pyhdfe.algorithm` to detect singleton fixed effects. It should be possible to greatly speed up this part  by porting the `pyhdfe` implementation to `numba`
- Time to demean: `14.1` - fairly fast but not `fixest` level fast. One easy to implement trick that `fixest` does but currently is not implemented is to sort the fixed effects by their weights prior to demeaning. Additional performance improvements are possible to implement a special logic for non-weighted demeaning. 

So I think that pushing performance to less than 1 minute for this problem should be very much achievable!

**Comment by s3alfisc** [comment]
Based on the [dev branch](https://github.com/s3alfisc/pyfixest/pull/191), we're now at 35 seconds or 3x of `fixest` for run 2 (after jit compilation). Demeaning takes around 11 seconds, hence 1x of `fixest` =)

```py
%timeit
feols_reg = feols(
    fml="tip_amount ~ trip_distance + passenger_count | dofw + vendor_id + payment_type", 
    data=nyc, vcov="iid"
)
# 35.08s
```

**Comment by s3alfisc** [comment]
Hi all, just a brief update - I merged a beta version of pyfixest `0.19` with many small performance tweaks and now `pyfixest.feols` solves Grant's NY taxi problem in around 18 seconds (though I am cheating a bit, as I am doing a very costly pandas object to categorical transform outside of `feols`): 

```python
%load_ext autoreload
%autoreload 2

import duckdb
import time
import numpy as np
import pyfixest as pf

# %%
nyc = duckdb.sql(
    '''
    FROM 'C:/Users/alexa/Documents/nyc-taxi/**/*.parquet'
    SELECT
        tip_amount, trip_distance, passenger_count,
        vendor_id, payment_type, dropoff_at,
        dayofweek(dropoff_at) AS dofw
    WHERE year = 2012 AND month <= 3
    '''
    ).df()

# convert dowf, vendor_id, payment_type to categorical
tic = time.time()
nyc["dofw"] = nyc["dofw"].astype(int)
nyc["vendor_id"] = nyc["vendor_id"].astype("category")
nyc["payment_type"] = nyc["payment_type"].astype("category")
print(f"""
    I am convering columns of type 'objects' to 'categories' and 'int'data types outside 
    of the regression, hence I am cheating a bit. This saves {np.round(time.time() - tic)} seconds.
    """
)
#    I am convering columns of type 'objects' to 'categories' and 'int'data types outside 
#    of the regression, hence I am cheating a bit. This saves 7.0 seconds.

run = True
if run: 

    # mock regression for JIT compilation
    fit = pf.feols(
        fml = "tip_amount ~ trip_distance + passenger_count | vendor_id + payment_type + dofw", 
        data = nyc.iloc[1:10_000], 
        copy_data = False,
        store_data = False
        )

    import time
    tic = time.time()
    fit = pf.feols(
        fml = "tip_amount ~ trip_distance + passenger_count | vendor_id + payment_type + dofw", 
        data = nyc, 
        copy_data = False, # saves a few seconds
        store_data = False # saves a few second
        )
    passed = time.time() - tic
    print(f"Passed time is {np.round(passed)}.")
    # Passed time is 22.
```

**Comment by grantmcdermott** [comment]
Super job Alex, nicely done!

Although you'll never convince me to use pandas ;-)
